pkg/cli/set/data.go

```diff
@@ -0,0 +1,443 @@
+		return err
+	}
+	o.Builder = f.NewBuilder
+	o.UpdateDataForObject = updateDataForObject
+
+	o.DryRun = kcmdutil.GetDryRunFlag(cmd)
+	if o.DryRun {
+		o.PrintFlags.Complete("%s (dry run)")
+	}
+	o.Printer, err = o.PrintFlags.ToPrinter()
+	if err != nil {
+		return err
+	}
+
+	clientConfig, err := f.ToRESTConfig()
+	if err != nil {
+		return err
+	}
+	o.Client, err = dynamic.NewForConfig(clientConfig)
+	if err != nil {
+		return err
+	}
+
+	args, dataArgs, err := kcmdutil.GetResourcesAndPairs(args, "data")
+	if err != nil {
+		return err
+	}
+
+	o.Resources = args
+
+	o.SetData, o.RemoveData, err = kcmdutil.ParsePairs(dataArgs, "data", true)
+	if err != nil {
+		return err
+	}
+
+	kv, err := keyValuesFromFileSources(o.FileSources)
+	if err != nil {
+		return err
+	}
+	for k, v := range kv {
+		if _, ok := o.SetData[k]; ok && k != v {
+			return fmt.Errorf("cannot set key %q in both argument and flag", k)
+		}
+		o.SetData[k] = v
+	}
+	kv, err = keyValuesFromFileSources(o.LiteralSources)
+	if err != nil {
+		return err
+	}
+	for k, v := range kv {
+		if _, ok := o.SetData[k]; ok && k != v {
+			return fmt.Errorf("cannot set key %q in both argument and flag", k)
+		}
+		o.SetData[k] = v
+	}
+
+	return nil
+}
+
+func (o *DataOptions) Validate() error {
+	if len(o.SetData) == 0 && len(o.RemoveData) == 0 {
+		return fmt.Errorf("must add, update, or remove at least one key from this object")
+	}
+	for _, remove := range o.RemoveData {
+		if _, ok := o.SetData[remove]; ok {
+			return fmt.Errorf("you may not remove and set the key %q in the same invocation", remove)
+		}
+	}
+	return nil
+}
+
+func (o *DataOptions) Run() error {
+	b := o.Builder().
+		WithScheme(scheme.Scheme, scheme.Scheme.PrioritizedVersionsAllGroups()...).
+		LocalParam(o.Local).
+		ContinueOnError().
+		NamespaceParam(o.Namespace).DefaultNamespace().
+		FilenameParam(o.ExplicitNamespace, &o.FilenameOptions).
+		Flatten()
+
+	if !o.Local {
+		b = b.
+			LabelSelectorParam(o.Selector).
+			ResourceTypeOrNameArgs(o.All, o.Resources...).
+			Latest()
+	}
+
+	singleItemImplied := false
+	infos, err := b.Do().IntoSingleItemImplied(&singleItemImplied).Infos()
+	if err != nil {
+		return err
+	}
+
+	allErrs := []error{}
+
+	patches := CalculatePatchesExternal(infos, func(info *resource.Info) (bool, error) {
+		changed := false
+		valid, err := o.UpdateDataForObject(info.Object, func(data map[string][]byte) error {
+			for k, v := range o.SetData {
+				if existing, ok := data[k]; ok && string(existing) == v {
+					continue
+				}
+				data[k] = []byte(v)
+				changed = true
+			}
+			for _, k := range o.RemoveData {
+				if _, ok := data[k]; !ok {
+					continue
+				}
+				delete(data, k)
+				changed = true
+			}
+			return nil
+		})
+		if valid && !changed {
+			if o.Local || o.DryRun {
+				if err := o.Printer.PrintObj(info.Object, o.Out); err != nil {
+					allErrs = append(allErrs, err)
+				}
+			} else {
+				fmt.Fprintf(o.ErrOut, "info: %s was not changed\n", info.Name)
```

Comment:
We don't print those explicitly anywhere else, only with `V(1)` just like below. With this you might end up with 2 infos about not changed, though.